Removing chain dependencies from NewCommitServices construct

patrickhuie19 opened the pull request:
## Motivation
LOOP-ify

## Solution

patrickhuie19 committed:
WIP; moved DynamicPriceGetterClient initialization to delegate from NewCommitServices

patrickhuie19 committed:
WIP; contract readers added as field to DynamicPriceGetter and sketched out ContractReader config

patrickhuie19 committed:
fix test usage of NewDynamicPriceGetter

patrickhuie19 committed:
WIP: added contract reader config and bindings

patrickhuie19 committed:
added debug image for smoke tests with delve and minor WIP steps towards using ChainReader

patrickhuie19 committed:
WIP use of CR for decimals results

Tofel committed:
expose delve port

patrickhuie19 committed:
fixing bug in build_ccip_debug_image make command naming

patrickhuie19 committed:
Adding install-chainlink-delve make command

patrickhuie19 committed:
minor fix to last commit

patrickhuie19 committed:
updated debug Dockerfile with delve installed in final image

Tofel committed:
run cl node container with dlv

Tofel committed:
use full dlv path

Tofel committed:
try another path

Tofel committed:
chainlink starts with dlv

patrickhuie19 committed:
changing final image for debug Dockerfile to golang:1.22-bullseye

patrickhuie19 committed:
udpated TestDynamicPriceGetterWithEmptyInput to use CR mocks

patrickhuie19 committed:
using GetEndpointFromPort in integration tests

patrickhuie19 committed:
fixed panic in Test_CLOSpecApprovalFlow_dynamicPriceGetter

patrickhuie19 committed:
bumping CTF

patrickhuie19 committed:
bumping go.mod

patrickhuie19 committed:
Merge branch 'ccip-develop' into feature/commit-init-no-evm

patrickhuie19 committed:
undoing mod bumps

patrickhuie19 committed:
bumping only CTF in integration-tests module

patrickhuie19 committed:
Test_CLOSpecApprovalFlow_dynamicPriceGetter passes

patrickhuie19 committed:
fixed TestDynamicPriceGetterWithEmptyInput

patrickhuie19 committed:
Merge branch 'ccip-develop' into feature/commit-init-no-evm

patrickhuie19 committed:
updating to use LatestRound CR + lint

patrickhuie19 committed:
reverting entrypoint to not use dlv binary

patrickhuie19 committed:
fixed overwriting of results struct by ChainReader, identified potential bug in how contract addresses are drilled to ChainReader

patrickhuie19 committed:
fixed conflicting contract names with indexed bindings. confirmed non-determinism in chain reader

patrickhuie19 committed:
fixed map iteration bug in Test_CLOSpecApprovalFlow_dynamicPriceGetter

patrickhuie19 committed:
fixed TestDynamicPriceGetterWithEmptyInput again

patrickhuie19 committed:
cleaned up hanging legacy evm batch call code from evm pricegetter'

patrickhuie19 committed:
cleaned up legacy clients in newServicesCommit

dimkouv commented on core/chainlink.debug.Dockerfile:
```diff
@@ -0,0 +1,97 @@
+# Build image: Chainlink binary
```
Why we have that?

patrickhuie19 replied:
This dockerfile allows a remote debugger (i.e. delve in GoLand or in your terminal) to connect to a node spun up with this Dockerfile. The application of this I was interested in was remote debugging into a node in the bidirectional CCIP smoke test

winder replied:
Neat. Could you put some comments at the top with some information about how to use it?

dimkouv commented on core/services/ocr2/delegate.go:
```diff
@@ -1618,7 +1616,80 @@ func (d *Delegate) newServicesCCIPCommit(ctx context.Context, lggr logger.Sugare
+			relayID := types.RelayID{Network: spec.Relay, ChainID: strconv.FormatUint(chainID, 10)}
+			relay, rerr := d.RelayGetter.Get(relayID)
+			if rerr != nil {
+				return nil, rerr
```
Wrap err e.g. `"get relay by id=%d: %w", err`

dimkouv commented on core/services/ocr2/delegate.go:
```diff
@@ -1618,7 +1616,80 @@ func (d *Delegate) newServicesCCIPCommit(ctx context.Context, lggr logger.Sugare
+				Contracts: map[string]evmrelaytypes.ChainContractReader{
+					// TODO CCIP-3444 can bind multiple contract addresses to the same contract name after merge,
+					// removing the need to have indexed contract names
+					"OffchainAggregator_0": {
```
Probably need to pull this in some const `OffchainAggregator_0` and re-use it while making the query

patrickhuie19 replied:
It's a little tougher to maneuver than i thought because the consts naturally would live in the evm price getter implementation (or a contracts dir that lives outside of core). However, its hidden behind an internal directory path. So for now they are duplicated, but this duplication will be stripped out in the v3 chainlink node

dimkouv commented on core/services/ocr2/delegate.go:
```diff
@@ -1618,7 +1616,80 @@ func (d *Delegate) newServicesCCIPCommit(ctx context.Context, lggr logger.Sugare
+					"OffchainAggregator_0": {
+						ContractABI: ccip.OffChainAggregatorABI,
+						Configs: map[string]*evmrelaytypes.ChainReaderDefinition{
+							"decimals": { // CR consumers choose an alias
```
Same for method names, etc.

dimkouv commented on core/services/ocr2/delegate.go:
```diff
@@ -1618,7 +1616,80 @@ func (d *Delegate) newServicesCCIPCommit(ctx context.Context, lggr logger.Sugare
+
+			contractReaderConfigJsonBytes, err := json.Marshal(contractReaderConfig)
+			if err != nil {
+				return nil, err
```
Wrap err `"marshal contract reader cfg: %w", err`

dimkouv commented on core/services/ocr2/delegate.go:
```diff
@@ -1618,7 +1616,80 @@ func (d *Delegate) newServicesCCIPCommit(ctx context.Context, lggr logger.Sugare
+
+			contractReader, err := relay.NewContractReader(ctx, contractReaderConfigJsonBytes)
+			if err != nil {
+				return nil, err
```
same

patrickhuie19 committed:
minor fix

patrickhuie19 committed:
lint

patrickhuie19 committed:
lint

patrickhuie19 commented on core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go:
```diff
@@ -21,6 +27,7 @@ import (
 
 const decimalsMethodName = "decimals"
 const latestRoundDataMethodName = "latestRoundData"
+const OFFCHAIN_AGGREGATOR = "OffchainAggregator"
```
We have mixed conventions for defining constants, but I can follow snakeCase if preferred

dimkouv replied:
imo minor

patrickhuie19 commented on core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go:
```diff
@@ -144,60 +151,101 @@ func (d *DynamicPriceGetter) performBatchCalls(ctx context.Context, batchCallsPe
+	for i, call := range batchCalls.decimalCalls {
+		bindings = append(bindings, types.BoundContract{
+			Address: string(ccipcalc.EvmAddrToGeneric(call.ContractAddress())),
+			Name:    fmt.Sprintf("%v_%v", OFFCHAIN_AGGREGATOR, i),
```
Once we merge back, this can simplified to use multiple contract addresses per bound contract name. (CCIP-3444)

patrickhuie19 committed:
clarifying comments

patrickhuie19 commented on core/services/ocr2/plugins/ccip/internal/pricegetter/evm_test.go:
```diff
@@ -760,12 +822,54 @@ func mockCaller(t *testing.T, decimals []uint8, rounds []aggregator_v3_interface
+		bGLVR[fmt.Sprintf("%v_%v", OFFCHAIN_AGGREGATOR, i)] = make([]types.BatchReadResult, 0, 2)
+	}
+	for i, d := range decimals {
+		contractName := fmt.Sprintf("%v_%v", OFFCHAIN_AGGREGATOR, i)
```
Can also be simplified once merged into chainlink. (CCIP-3444)

patrickhuie19 committed:
cleaning up delve entrypoint integ tests and minor refactoring for clarity

patrickhuie19 commented on core/services/ocr2/delegate.go:
```diff
@@ -1618,7 +1616,80 @@ func (d *Delegate) newServicesCCIPCommit(ctx context.Context, lggr logger.Sugare
+				Contracts: map[string]evmrelaytypes.ChainContractReader{
+					// TODO CCIP-3444 can bind multiple contract addresses to the same contract name after merge,
+					// removing the need to have indexed contract names
+					"OffchainAggregator_0": {
```
This should be improved if possible by tying it into information passed to the job - can we determine how many contract addresses in a given chain `TokenPricesUSD` or other consumers of the dynamic price getter may use directly from the job spec? 

@dimkouv

dimkouv replied:
You mean putting in the jobspec the abi?

patrickhuie19 replied:
Fixed in https://github.com/smartcontractkit/ccip/commit/0b5b79d104386697817ef5635c13aae89e7a73f8

winder commented on core/services/ocr2/delegate.go:
```diff
@@ -1618,7 +1616,80 @@ func (d *Delegate) newServicesCCIPCommit(ctx context.Context, lggr logger.Sugare
+				return nil, rerr
+			}
+
+			contractReaderConfig := evmrelaytypes.ChainReaderConfig{
+				Contracts: map[string]evmrelaytypes.ChainContractReader{
+					// TODO CCIP-3444 can bind multiple contract addresses to the same contract name after merge,
+					// removing the need to have indexed contract names
+					"OffchainAggregator_0": {
+						ContractABI: ccip.OffChainAggregatorABI,
+						Configs: map[string]*evmrelaytypes.ChainReaderDefinition{
+							"decimals": { // CR consumers choose an alias
+								ChainSpecificName: "decimals",
+							},
+							"latestRoundData": {
+								ChainSpecificName: "latestRoundData",
+							},
+						},
+					},
+					"OffchainAggregator_1": {
+						ContractABI: ccip.OffChainAggregatorABI,
+						Configs: map[string]*evmrelaytypes.ChainReaderDefinition{
+							"decimals": { // CR consumers choose an alias
+								ChainSpecificName: "decimals",
+							},
+							"latestRoundData": {
+								ChainSpecificName: "latestRoundData",
+							},
+						},
+					},
+				},
+			}
```
```suggestion
			aggregatorConfig := evmrelaytypes.ChainContractReader {
				ContractABI: ccip.OffChainAggregatorABI,
				Configs: map[string]*evmrelaytypes.ChainReaderDefinition{
					"decimals": { // CR consumers choose an alias
						ChainSpecificName: "decimals",
					},
					"latestRoundData": {
						ChainSpecificName: "latestRoundData",
					},
				},
			},
			
			contractReaderConfig := evmrelaytypes.ChainReaderConfig{
				Contracts: map[string]evmrelaytypes.ChainContractReader{
					// TODO CCIP-3444 can bind multiple contract addresses to the same contract name after merge,
					// removing the need to have indexed contract names
					"OffchainAggregator_0": aggregatorConfig,
					"OffchainAggregator_1": aggregatorConfig,
				},
			}
```

patrickhuie19 replied:
Thanks :) I can do even better since discussing https://github.com/smartcontractkit/ccip/pull/1361#discussion_r1765428196 offline with Dim. I'll post a revision soon

patrickhuie19 replied:
https://github.com/smartcontractkit/ccip/pull/1361/commits/0b5b79d104386697817ef5635c13aae89e7a73f8

patrickhuie19 committed:
using job spec to determine the number of contract addresses per chain contract reader must be configured for

patrickhuie19 committed:
wrapping errors, exporting OffchainAggregator metadata vars

patrickhuie19 committed:
Merge branch 'ccip-develop' into feature/commit-init-no-evm

patrickhuie19 committed:
fixing import cycle

dimkouv reviewed (commented):
let's wait for flaky tests to be resolved before merging

patrickhuie19 commented:
> let's wait for flaky tests to be resolved before merging

@Tofel @lukaszcl re: testcontainers bug

lukaszcl committed:
Try different runner for e2e test

lukaszcl committed:
Revert "Try different runner for e2e test"

This reverts commit b323a8b1ab6270ab602df205c0564f2c0cfb46ab.

lukaszcl committed:
Feature/commit init no evm lf tests (#1470)

This PR tests https://github.com/smartcontractkit/ccip/pull/1361/files
without changes made to integration-tests/docker/test_env/cl_node.go

---------

Co-authored-by: Patrick <patrick.huie@smartcontract.com>

patrickhuie19 committed:
Merge branch 'ccip-develop' into feature/commit-init-no-evm

asoliman92 commented on core/services/ocr2/plugins/ccip/internal/pricegetter/evm_test.go:
```diff
@@ -200,6 +206,12 @@ func testParamAggregatorOnly(t *testing.T) testParameters {
 		uint64(103): mockClient(t, []uint8{18}, []aggregator_v3_interface.LatestRoundData{round3}),
 		uint64(104): mockClient(t, []uint8{20}, []aggregator_v3_interface.LatestRoundData{round4}),
 	}
+	contractReaders := map[uint64]types.ContractReader{
```
Do we still need this?
```
evmClients := map[uint64]DynamicPriceGetterClient{
		uint64(101): mockClient(t, []uint8{8}, []aggregator_v3_interface.LatestRoundData{round1}),
		uint64(102): mockClient(t, []uint8{8}, []aggregator_v3_interface.LatestRoundData{round2}),
		uint64(103): mockClient(t, []uint8{18}, []aggregator_v3_interface.LatestRoundData{round3}),
		uint64(104): mockClient(t, []uint8{20}, []aggregator_v3_interface.LatestRoundData{round4}),
	}
```

patrickhuie19 replied:
addressed in https://github.com/smartcontractkit/ccip/pull/1472/files

winder commented on core/services/ocr2/delegate.go:
```diff
@@ -1618,7 +1616,77 @@ func (d *Delegate) newServicesCCIPCommit(ctx context.Context, lggr logger.Sugare
+			if rerr != nil {
+				return nil, fmt.Errorf("get relay by id=%v: %w", relayID, err)
+			}
+
+			contractsConfig := make(map[string]evmrelaytypes.ChainContractReader, len(aggregatorContracts))
+			for i := range aggregatorContracts {
+				contractsConfig[fmt.Sprintf("%v_%v", ccip.OFFCHAIN_AGGREGATOR, i)] = evmrelaytypes.ChainContractReader{
+					ContractABI: ccip.OffChainAggregatorABI,
+					Configs: map[string]*evmrelaytypes.ChainReaderDefinition{
+						"decimals": { // CR consumers choose an alias
+							ChainSpecificName: "decimals",
+						},
+						"latestRoundData": {
+							ChainSpecificName: "latestRoundData",
+						},
+					},
+				}
+			}
+			contractReaderConfig := evmrelaytypes.ChainReaderConfig{
+				Contracts: contractsConfig,
+			}
+
+			contractReaderConfigJsonBytes, jerr := json.Marshal(contractReaderConfig)
+			if jerr != nil {
+				return nil, fmt.Errorf("marshal contract reader config: %w", jerr)
+			}
+
+			contractReader, cerr := relay.NewContractReader(ctx, contractReaderConfigJsonBytes)
+			if cerr != nil {
+				return nil, fmt.Errorf("new ccip commit contract reader %w", cerr)
+			}
+
+			contractReaders[chainID] = contractReader
+		}
+
+		priceGetter, err = ccip.NewDynamicPriceGetter(*pluginJobSpecConfig.PriceGetterConfig, contractReaders)
+		if err != nil {
+			return nil, fmt.Errorf("creating dynamic price getter: %w", err)
+		}
+	}
```
It would be nice to wrap this in a function similar to `d.ccipCommitGetDstProvider` and `d.ccipCommitGetSrcProvider` above.

patrickhuie19 replied:
addressed in https://github.com/smartcontractkit/ccip/pull/1472/files

winder commented on core/services/ocr2/plugins/ccip/internal/pricegetter/evm.go:
```diff
@@ -7,29 +7,31 @@ import (
 
-const decimalsMethodName = "decimals"
-const latestRoundDataMethodName = "latestRoundData"
+const OFFCHAIN_AGGREGATOR = "OffchainAggregator"
+const DECIMALS_METHOD_NAME = "decimals"
+const LATEST_ROUND_DATA_METHOD_NAME = "latestRoundData"
```
All caps is not idiomatic (stdlib uses UpperCamelCase, i.e. `math.MaxUint64`). They were also private before. So two requests:
1. Can it be private?
2. Please rename to `DecimalsMethodName` or `decimalsMethodName`

patrickhuie19 replied:
These specific vars are not exported anywhere else, since they live in a package that is defined behind in a leaf of an`internal` path. Will rename to UpperCamelCase, thanks for the stdlib callout.

patrickhuie19 replied:
addressed in https://github.com/smartcontractkit/ccip/pull/1472/files

winder commented on GNUmakefile:
```diff
@@ -48,6 +48,10 @@ install-chainlink: operator-ui ## Install the chainlink binary.
 install-chainlink-cover: operator-ui ## Install the chainlink binary with cover flag.
 	go install -cover $(GOFLAGS) .
 
+.PHONY: install-chainlink-delve
+install-chainlink-delve: operator-ui ## Install the chainlink binary.
+	go install $(GOFLAGS) -gcflags "all=-N -l" .
```
What is this? It doesn't seem to have anything to do with delve

patrickhuie19 replied:
The `-gcflags "all=-N -l"` portion of this compiles the binary in a way that allows a debugger to step through.

winder reviewed (approved):
Approved, but hoping for a followup to address the nits

patrickhuie19 committed:
Merge branch 'ccip-develop' into feature/commit-init-no-evm
